coil-compose-core/src/commonMain/kotlin/coil3/compose/internal/ForwardingUnconfinedCoroutineScope.kt

```diff
@@ -0,0 +1,61 @@
+package coil3.compose.internal
+
+import kotlin.coroutines.CoroutineContext
+import kotlinx.atomicfu.atomic
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.InternalCoroutinesApi
+import kotlinx.coroutines.Runnable
+
+/**
+ * A [CoroutineScope] with a special [CoroutineContext] that enables [ForwardingUnconfinedCoroutineDispatcher]
+ * to enable dispatching after it is replaced in the context.
+ */
+internal fun ForwardingUnconfinedCoroutineScope(
+    context: CoroutineContext,
+) = CoroutineScope(
+    context = ForwardingCoroutineContext(context) { old, new ->
+        val oldDispatcher = old.dispatcher
+        val newDispatcher = new.dispatcher
+        if (oldDispatcher is ForwardingUnconfinedCoroutineDispatcher && oldDispatcher != newDispatcher) {
+            oldDispatcher.unconfined = oldDispatcher.unconfined &&
+                    (newDispatcher == null || newDispatcher == Dispatchers.Unconfined)
+        }
+    }
+)
+
+/**
+ * A [CoroutineDispatcher] that delegates to [Dispatchers.Unconfined] while [unconfined] is true
+ * and [delegate] when [unconfined] is false.
+ */
+internal class ForwardingUnconfinedCoroutineDispatcher(
+    private val delegate: CoroutineDispatcher,
+) : CoroutineDispatcher() {
+    private val _unconfined = atomic(true)
+    var unconfined by _unconfined
```

Comment:
Consider documenting that after setting this to false, it's ignored even if set to true again.